### PR TITLE
Implement detection for sti for including subclasses in check

### PR DESCRIPTION
**Pull request by Liberatys**
* Remove subclass rule inclusion for non sti classes
* Start of implementing more sti detection utilities which are going to be used in a rewrite (coming)

@coorasse how do you want to handle this ? It seems like STI will have some issues with the current implementation. I'll probably have to add conditional loading at 3 places to fully comply with the expected behaviour.

### CODE REVIEW

File: lib/cancan/class_matcher.rb [hunk 1]
```diff
@@ -12,6 +14,8 @@ def self.matches_subject_class?(subjects, subject)
   def self.matching_class_check(subject, sub, has_subclasses)
     matches = matches_class_or_is_related(subject, sub)
     if has_subclasses
+      return matches unless StiDetector.sti_class?(sub)
```
Comment (Liberatys) [inline lib/cancan/class_matcher.rb:17]: Only include the subclasses in the check if the model is an sti class

File: lib/cancan/sti_detector.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class StiDetector
+  def self.sti_class?(subject)
```
Comment (Liberatys) [inline lib/cancan/sti_detector.rb:2]: A general detection for STI classes

File: spec/cancan/model_adapters/active_record_adapter_spec.rb [hunk 3]
```diff
@@ -909,6 +909,37 @@ class JsonTransaction < ActiveRecord::Base
     end
   end
 
+  context 'with rule application to subclass for non sti class' do
```
Comment (Liberatys) [inline spec/cancan/model_adapters/active_record_adapter_spec.rb:912]: Check that rules are no longer applied to parents for non sti

File: spec/cancan/model_adapters/active_record_adapter_spec.rb [hunk 4]
```diff
@@ -909,6 +909,37 @@ class JsonTransaction < ActiveRecord::Base
...
+      end
+
+      class Parent < ActiveRecord::Base
+      end
+
+      class Child < Parent
+      end
+    end
+
+    it 'rules are not effecting parent class' do
```
Comment (phallstrom) [inline spec/cancan/model_adapters/active_record_adapter_spec.rb:933]: I think we also need the test where can'ing the child doesn't automatically can the parent to satisfy this scenario, yeah? https://github.com/CanCanCommunity/cancancan/issues/677#issuecomment-753387317

Something like...

```
u1 = User.create!(name: 'pippo')
ability = Ability.new(u1)
ability.can :manage, Child
expect(ability).to be_able_to(:index, Child)
expect(ability).not_to be_able_to(:index, Parent)
```

**Comment by winstonwolff** [comment]
Thank you for working on this. We've tried this code and it seems to work—our test cases pass and it seems to block a parent class of resources as we expected. If we see any problems with it, we'll report back.

**Comment by swelther** [comment]
Any news on this? 🙏 This issue prevents us from updating to cancancan 3.2+ :(

File: lib/cancan/sti_detector.rb (hunk 2 above)
Comment (coorasse) [reply lib/cancan/sti_detector.rb:4]: Love its isolation

**Review by coorasse — COMMENTED** [review]
@Liberatys do we need to check further scenarios here, like @phallstrom suggested, or are they covered already somewhere else?

**Comment by Liberatys** [comment]
@coorasse I now also added the spec that covers the example from #677. CI is blue/green ^^

**Review by coorasse — APPROVED** [review]

**Pull request merged**